osquery/utils/windows/shelllnk.cpp

```diff
@@ -0,0 +1,635 @@
+    std::string type = data.substr((offset * 2) + 32, 8);
+    std::string location_type = getLocationType(type);
+    if (location_type.empty()) {
+      data.erase(0, location_size);
+      location_info.data = data;
+      return location_info;
+    }
+    location_info.type = location_type;
+    std::string common_path_offset = data.substr(48, 8);
+    common_path_offset = swapEndianess(common_path_offset);
+    int path_offset = tryTo<int>(common_path_offset, 16).takeOr(0) * 2;
+    size_t common_path_size = data.find("00", path_offset);
+    // Size should be even but if values end in base 10 it will be odd
+    if (common_path_size % 2 != 0) {
+      common_path_size++;
+    }
+    std::string common_path =
+        data.substr(path_offset, common_path_size - path_offset);
+    try {
+      location_info.common_path = boost::algorithm::unhex(common_path);
+    } catch (const boost::algorithm::hex_decode_error& /* e */) {
+      LOG(WARNING) << "Failed to decode common path hex values to string: "
+                   << common_path;
+    }
+    std::string share_name_offset = data.substr((offset * 2) + 16, 8);
+    share_name_offset = swapEndianess(share_name_offset);
+    int share_offset = tryTo<int>(share_name_offset, 16).takeOr(0);
+    size_t share_name_size = data.find("00", (offset * 2) + share_offset * 2);
+    std::string share_name =
+        data.substr((offset * 2) + share_offset * 2,
+                    share_name_size - ((offset * 2) + share_offset * 2));
+    try {
+      location_info.share_name = boost::algorithm::unhex(share_name);
+    } catch (const boost::algorithm::hex_decode_error& /* e */) {
+      LOG(WARNING) << "Failed to decode share name hex values to string: "
+                   << share_name;
+    }
+  } else {
+    LOG(WARNING) << "Unknown location type: " << location_type;
+  }
+  data.erase(0, location_size);
+  location_info.data = data;
+  return location_info;
+}
+
+DataStringInfo parseDataString(const std::string& data,
+                               const bool unicode,
+                               const bool description,
+                               const bool relative_path,
+                               const bool working_path,
+                               const bool icon_location,
+                               const bool command_args) {
+  std::string data_string = data;
+  std::string data_str_type;
+  DataStringInfo data_info;
+  std::string str_data_size = data_string.substr(0, 4);
+  // Previous lnk data may have extra zeros (due to Unicode null termination?)
+  if (str_data_size == "0000") {
+    data_string.erase(0, 4);
+    str_data_size = data_string.substr(0, 4);
+  }
+  str_data_size = swapEndianess(str_data_size);
+  int data_size = tryTo<int>(str_data_size, 16).takeOr(0);
+
+  // Data strings go in the following order: description, relative path, working
+  // path, command args, icon location. Some may not exist
+  if (description) {
```

Comment:
It looks like this series (description, relative_path, working_path, etc) of checks and parsing are the same? Can you create a function for this and reduce the code repetition?

Reply:
good point, i made small function. The function just returns the expected string (description, relative_path, etc)
The only part that it does not do is calculate the size of the next data block which is still in the conditional statements

Reply:
You might be able to do something like:

```
auto dataStringParse = [&data_size, &data_string](std::string& target, bool unicode) {
  target = parseLnkData(data_string, unicode, data_size);
  if (unicode) {
    data_size = data_size * 4;
  }
  data_string.erase(0, data_size + 4);
  auto str_data_size = data_string.substr(0, 4);
  str_data_size = swapEndianess(str_data_size);
  data_size = tryTo<int>(str_data_size, 16).takeOr(0);
}

if (description) {
  dataStringParse(data_info.description, unicode);
}
if (relative_path) {
  dataStringParse(data_info.relative_path, unicode);
}
...
```

Reply:
thanks for this!
works perfectly